File: controllers/userController.js
import User from "../models/User.js";

/* ================= GET USER PROFILE ================= */
export const getUserProfile = async (req, res) => {
  try {
    const user = await User.findById(req.userId).select("-password");
    
    if (!user) {
      return res.status(404).json({ message: "User not found" });
    }

    res.json(user);
  } catch (error) {
    console.error("Get Profile Error:", error.message);
    res.status(500).json({ message: "Failed to fetch profile", error: error.message });
  }
};

/* ================= UPDATE USER PROFILE ================= */
export const updateUserProfile = async (req, res) => {
  try {
    console.log("=== UPDATE PROFILE REQUEST ===");
    console.log("User ID:", req.userId);
    console.log("Request Body:", req.body);

    const user = await User.findById(req.userId);
    
    if (!user) {
      console.log("User not found");
      return res.status(404).json({ message: "User not found" });
    }

    // Simple update - just copy values
    const { name, height, weight, age, gender, foodPreference, diseases } = req.body;

    if (name) user.name = name;
    if (age) user.age = Number(age);
    if (gender) user.gender = gender;
    if (foodPreference) user.foodPreference = foodPreference;
    
    // Handle height and weight
    if (height) user.height = Number(height);
    if (weight) user.weight = Number(weight);
    
    // Calculate BMI
    if (user.height && user.weight) {
      const h = user.height / 100;
      user.bmi = Number((user.weight / (h * h)).toFixed(1));
    }
    
    // Handle diseases
    if (diseases !== undefined) {
      if (Array.isArray(diseases)) {
        user.diseases = diseases;
      } else if (typeof diseases === 'string') {
        user.diseases = diseases ? diseases.split(',').map(d => d.trim()) : [];
      } else {
        user.diseases = [];
      }
    }

    console.log("Saving user...");
    await user.save();
    console.log("User saved successfully");

    const updatedUser = await User.findById(req.userId).select("-password");
    
    res.json({
      message: "Profile updated successfully",
      user: updatedUser
    });
  } catch (error) {
    console.error("=== UPDATE PROFILE ERROR ===");
    console.error("Error:", error);
    
    res.status(500).json({ 
      message: "Failed to update profile",
      error: error.message
    });
  }
};

/* ================= GET BMI HISTORY ================= */
export const getBMIHistory = async (req, res) => {
  try {
    const user = await User.findById(req.userId);
    
    if (!user) {
      return res.status(404).json({ message: "User not found" });
    }

    res.json({
      bmiHistory: user.bmiHistory || []
    });
  } catch (error) {
    console.error("Get BMI History Error:", error.message);
    res.status(500).json({ message: "Failed to fetch BMI history", error: error.message });
  }
};

/* ================= ADD BMI RECORD ================= */
export const addBMIRecord = async (req, res) => {
  try {
    const { weight, height, bmi } = req.body;

    const user = await User.findById(req.userId);
    
    if (!user) {
      return res.status(404).json({ message: "User not found" });
    }

    if (!user.bmiHistory) {
      user.bmiHistory = [];
    }

    user.bmiHistory.push({
      weight: Number(weight),
      height: Number(height),
      bmi: Number(bmi),
      date: new Date()
    });

    // Keep only last 30 records
    if (user.bmiHistory.length > 30) {
      user.bmiHistory = user.bmiHistory.slice(-30);
    }

    user.weight = Number(weight);
    user.height = Number(height);
    user.bmi = Number(bmi);

    await user.save();

    res.json({
      message: "BMI record added successfully",
      bmiHistory: user.bmiHistory
    });
  } catch (error) {
    console.error("Add BMI Record Error:", error.message);
    res.status(500).json({ message: "Failed to add BMI record", error: error.message });
  }
};

File: controllers/user_controller.py
from fastapi import HTTPException
from pydantic import BaseModel
from typing import Optional, List
from beanie import PydanticObjectId
from models.user import User, BMIRecord
from datetime import datetime


# ================= REQUEST SCHEMAS =================

class UpdateProfileRequest(BaseModel):
    name: Optional[str] = None
    height: Optional[float] = None
    weight: Optional[float] = None
    age: Optional[int] = None
    gender: Optional[str] = None
    foodPreference: Optional[str] = None
    diseases: Optional[object] = None  # Can be list or comma-separated string


class AddBMIRequest(BaseModel):
    weight: float
    height: float
    bmi: float


# ================= HELPER =================

def to_oid(user_id: str) -> PydanticObjectId:
    return PydanticObjectId(user_id)


# ================= GET USER PROFILE =================

async def get_user_profile(user_id: str):
    try:
        user = await User.get(to_oid(user_id))
        if not user:
            raise HTTPException(status_code=404, detail="User not found")

        user_data = user.dict()
        user_data["_id"] = str(user.id)
        user_data.pop("password", None)
        return user_data

    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to fetch profile: {str(e)}")


# ================= UPDATE USER PROFILE =================

async def update_user_profile(data: UpdateProfileRequest, user_id: str):
    try:
        user = await User.get(to_oid(user_id))
        if not user:
            raise HTTPException(status_code=404, detail="User not found")

        if data.name is not None:
            user.name = data.name
        if data.age is not None:
            user.age = int(data.age)
        if data.gender is not None:
            user.gender = data.gender
        if data.foodPreference is not None:
            user.foodPreference = data.foodPreference
        if data.height is not None:
            user.height = float(data.height)
        if data.weight is not None:
            user.weight = float(data.weight)

        # BMI calculate karo
        if user.height and user.weight:
            h = user.height / 100
            user.bmi = round(user.weight / (h * h), 1)

        # Diseases handle karo
        if data.diseases is not None:
            if isinstance(data.diseases, list):
                user.diseases = data.diseases
            elif isinstance(data.diseases, str):
                user.diseases = [d.strip() for d in data.diseases.split(",")] if data.diseases else []
            else:
                user.diseases = []

        user.updatedAt = datetime.now()
        await user.save()

        updated_user = user.dict()
        updated_user["_id"] = str(user.id)
        updated_user.pop("password", None)

        return {"message": "Profile updated successfully", "user": updated_user}

    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to update profile: {str(e)}")


# ================= GET BMI HISTORY =================

async def get_bmi_history(user_id: str):
    try:
        user = await User.get(to_oid(user_id))
        if not user:
            raise HTTPException(status_code=404, detail="User not found")

        return {"bmiHistory": user.bmiHistory or []}

    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to fetch BMI history: {str(e)}")


# ================= ADD BMI RECORD =================

async def add_bmi_record(data: AddBMIRequest, user_id: str):
    try:
        user = await User.get(to_oid(user_id))
        if not user:
            raise HTTPException(status_code=404, detail="User not found")

        if user.bmiHistory is None:
            user.bmiHistory = []

        user.bmiHistory.append(BMIRecord(
            weight=float(data.weight),
            height=float(data.height),
            bmi=float(data.bmi),
            date=datetime.now()
        ))

        # Last 30 records hi rakhna
        if len(user.bmiHistory) > 30:
            user.bmiHistory = user.bmiHistory[-30:]

        user.weight = float(data.weight)
        user.height = float(data.height)
        user.bmi = float(data.bmi)
        user.updatedAt = datetime.now()

        await user.save()

        return {
            "message": "BMI record added successfully",
            "bmiHistory": [r.dict() for r in user.bmiHistory]
        }

    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to add BMI record: {str(e)}")
